### PR TITLE
Fixed various bugs in last_insert_id

### PR DESCRIPTION
**Pull request by zachmu**
Fixes https://github.com/dolthub/dolt/issues/6776

Also adds better test coverage for `LAST_INSERT_ID()` and the `INSERT_ID` field in the ok response for updates.

### CODE REVIEW

File: sql/rowexec/dml_iters.go [hunk 1]
```diff
@@ -516,29 +524,35 @@ func (a *accumulatorIter) Next(ctx *sql.Context) (r sql.Row, err error) {
...
 				ctx.Session.SetLastQueryInfo(sql.LastInsertId, oldLastInsertId)
 			}
 
+			res := a.updateRowHandler.okResult() // TODO: Should add warnings here
+
+			// For some update accumulators, we don't accurately track the last insert ID in the handler and need to set
+			// it manually in the result by getting it from the session. This doesn't work correctly in all cases and needs
+			// to be fixed. See comment in buildRowUpdateAccumulator in rowexec/dml.go
+			switch a.updateRowHandler.(type) {
+			case *onDuplicateUpdateHandler, *replaceRowHandler:
```
Comment (max-hoffman) [inline sql/rowexec/dml_iters.go:549]: not sure i follow what's special about these

**Review by max-hoffman — APPROVED** [review]
Seems like it fixes the customer bug. Should we be worried about `onDuplicateUpdateHandler` and `replaceRowHandler` special cases? Some of the mysql examples also get a bit more complicated, with setting expressions and stuff.

File: sql/rowexec/dml_iters.go (hunk 1 above)
Comment (zachmu) [reply sql/rowexec/dml_iters.go:549]: I'll update the comment.

They're accumulators that can be put into place for insert statements

**Comment by zachmu** [comment]
> Seems like it fixes the customer bug. Should we be worried about `onDuplicateUpdateHandler` and `replaceRowHandler` special cases? Some of the mysql examples also get a bit more complicated, with setting expressions and stuff.

Those two have the same buggy behavior as before, so the reporting of `last_insert_id` in the response won't be correct for `on duplicate key update` and `replace into` statements.

**Pull request merged**